File: rm_control/src/module/bullet_cover.cpp
#include <robot_toolbox/tool.h>
#include <robot_msgs/BoolStamped.h>

#include "rm_control/module/bullet_cover.h"

namespace rm_control
{
BulletCoverModule::BulletCoverModule(ros::NodeHandle& node, ros::NodeHandle& nodeParam)
    : ModuleInterface(node, nodeParam_), node_(node), nodeParam_(nodeParam)
{
}

bool BulletCoverModule::init()
{
    /* 读取配置 */
    CONFIG_ASSERT("bullet_cover/command_topic", nodeParam_.getParam("bullet_cover/command_topic", commandTopic_));
    commandTopicPublisher_ = node_.advertise<robot_msgs::BoolStamped>(commandTopic_, 1000, false);
    return true;
}

void BulletCoverModule::getValue(double& vx, double& vy, double& vrz, double& yawAngle, double& pitchAngle, ShotStatus& shotStatus, bool& isEnable, ros::Duration period)
{
    robot_msgs::BoolStamped msg;
    msg.header.seq++;
    msg.header.stamp = ros::Time::now();
    msg.result = isEnable;
    commandTopicPublisher_.publish(msg);
}
}  // namespace rm_control

File: rm_control_infantry/src/rm_control_infantry/modules/bullet_cover.py
from ..module import ModuleBase
from robot_msgs.msg import BoolStamped
from robot_toolbox.tools import ConfigAssert, CheckValidKey
from rm_referee_controller.msg import GameStatus
import rospy

class BulletCoverModule(ModuleBase):
    def __init__(self):
        super().__init__()
        self.__cmdMsg = BoolStamped()
        self.isOpen, self.__lastToggleKey = False, False

    def init(self):
        self.__toggleKey = rospy.get_param("~bullet_cover/keyboard/toggle")
        if not ConfigAssert("bullet_cover/keyboard/toggle", CheckValidKey(self.__toggleKey)):
            return False
        self.__cmdPublisher = rospy.Publisher(rospy.get_param("~bullet_cover/topic/cmd"), BoolStamped, queue_size=1000)
        return True
    
    def run(self, main, rc, mouse, keyboard, joints):
        isKeyPress = getattr(keyboard, self.__toggleKey)
        if isKeyPress != self.__lastToggleKey and isKeyPress:
            self.isOpen = not self.isOpen
        self.__cmdMsg.result = self.isOpen
        # 发送话题
        self.__cmdMsg.header.seq += 1
        self.__cmdMsg.header.stamp = rospy.Time.now()
        self.__cmdPublisher.publish(self.__cmdMsg)